Use java.lang.reflect.Type instead of Class<?> in factories

qualidafial opened the pull request:
## To do

* [x] Move ArgumentFactory to TypeToken
* [x] Ditto ResultColumnMapperFactory
* [x] Ditto ResultSetMapperFactory
* [x] Ditto CollectorFactory
  * [x] Change ResultBearing.collectInto(Class<T>) to TypeToken<T>?
* [x] Keep all the method that take Class<T>, and add a sister method that accepts TypeToken<T>
  * [x] SQLStatement.dynamicBind
* [x] Expose argument factory registry (Foreman) through statement context.
* [x] Enhance existing BinderFactory implementations to take advanced of generic type information.
* [x] Implement OptionalArgument
* [x] Confirm ResultReturnThing is getting full generic type information
* [x] Use Classmate GenericType<T> instead of Guava TypeToken<T>

## Questions

* Change SQLObject from Class<T> to TypeToken<T>? Implies inspecting a possibly generified DAO class and producing methods depending on the nested parameters in the type token.
  * Handle.attach
  * DBI.open
  * DBI.onDemand
  * SqlObjectBuilderBridge methods
  * StatementContext.sqlObjectType property
  * SqlStatementCustomer methods
  * CustomizingStatementHandler ctor parameter
  * CreateSqlObjectHandler
  * SqlObject methods
  * SqlObjectBuilder methods
  * SqlStatementCustomizingFactory methods
  * StringTemplate3StatementLocator?

* Built in row mappers use TypeToken instead of Class? Implies inspecting a possibly generified bean class and projecting that type onto generic members.
  * BeanMapper
    * BindBean can't do the converse--all we have is the instance class with generics erased.
  * FieldMapper
  * JpaMapper
  * Query.mapToBean(TypeToken<T>)

* Nothing we're doing here can be ported to existing annotations with methods that return Class types.

## Unrelated todos discovered while working on this--not until after this PR

* Rename registry methods `xyzFor` to `findXyzFor` and return an `Optional`
* StatementContext.columnMapperFor renamed to findColumnMapperFor, returns Optional
* Collapse factory `accepts` and `build` methods into a single `provide` (or otherwise named) method that returns an `Optional`.
* Remove generic type parameter from `ArgumentFactory`?
* [x] Remove generic type parameter from `CollectorFactory`?
* [x] Remove ResultBearing.collectInto(Class)--it's broken
  * We could possibly do collectInto(TypeToken<T>), but CollectorFactory would have to be verify the element type agrees with the collector result type.
* [x] BuiltInArgumentFactory enum encoding should use `Enum::name`
* Rename built in column mappers to exclude `Column` from name--it's redundant
* Do we need all those SQLStatement.bind(int|String, SpecificType) methods? Seems like a design smell to have these for every built in type, but users are on their own for custom types? 
* Eradicate `foo != null ? foo.something() : somethingElse`
* Likewise for `if (!x) { foo(); } else { bar(); }`
* `dynamicBind` should just overload `bind` and put the argument type last, effectively as an optional parameter
* Should we get rid of ObjectArgument? SQLStatement.bind(int|String, Object)? Or maybe require users to call overloaded method with type parameter if the argument might be null
  * If so, should also get rid of Foreman.waffle fallback to argument of type Object. An argument factory should be provided.
* SqlObjectBuilderBridge: Should this be a ServiceLoader SPI instead of using reflection?
* SqlStatementCustomizingAnnotation should be annotated `@Target(ElementType.ANNOTATION_TYPE)`
* `@MapResultAsBean` is equivalent to `@RegisterMapperFactory(BeanMapperFactory.class)`. Do we need both?
* `@Mapper` should be a SqlStatementCustomizingAnnotation instead of magic.
* Should EnumColumnMapper be exploded into EnumByNameColumnMapper and EnumByOrdinalColumnMapper?
* `parameter` vs `paramater`
* Binder.bind needs a resolved parameter type argument to address #242

qualidafial commented:
Idea: for collector factories, provide a method `findElementType(TypeToken<CollectionType>)` which if you passed e.g. a `TypeToken<Set<Foo>>` to the Set collector factory, would return a `TypeToken<Foo>`, thus allowing a query to infer both `mapTo` and `collect` steps into one method call:

    return getHandle().createQuery("select * from foo")
        .collectInto(new TypeToken<Set<Foo>>() {});

qualidafial commented:
@brianm @stevenschlansker This is ready for review.

Please take a look at the "questions" section in the PR description, and let me know your thoughts.

I've tried to keep this branch tightly focused--hence the "unrelated todos" section.

I'm not yet convinced this change is worth it.

/cc #205 #226

brianm commented:
Does this add a dependency on Guava? If so I think that is a bad idea. Gave breaks compact on releases, besides which jdbi has no runtime deps right now. As much as I like guava, it is dodgy for libraries to depend on it :-(

> On Nov 27, 2015, at 10:24 PM, Matthew Hall <notifications@github.com> wrote:
> 
> @brianm @stevenschlansker This is ready for review.
> 
> Please take a look at the "questions" section in the PR description, and let me know your thoughts.
> 
> I've tried to keep this branch tightly focused--hence the "unrelated todos" section.
> 
> I'm not yet convinced this change is worth it.
> 
> /cc #205 #226
> 
> ―
> Reply to this email directly or view it on GitHub.
>

qualidafial commented:
Yeah, and TypeToken is also `@Beta`. Since Guava is Apache 2 licensed, I
was thinking of just copying the class to JDBI.

Then I could add some methods to get the generic type parameters off the
TypeToken directly instead of through the java.lang.reflect API.
On Nov 28, 2015 08:48, "Brian McCallister" <notifications@github.com> wrote:

> Does this add a dependency on Guava? If so I think that is a bad idea.
> Gave breaks compact on releases, besides which jdbi has no runtime deps
> right now. As much as I like guava, it is dodgy for libraries to depend on
> it :-(
>
> > On Nov 27, 2015, at 10:24 PM, Matthew Hall <notifications@github.com>
> wrote:
> >
> > @brianm @stevenschlansker This is ready for review.
> >
> > Please take a look at the "questions" section in the PR description, and
> let me know your thoughts.
> >
> > I've tried to keep this branch tightly focused--hence the "unrelated
> todos" section.
> >
> > I'm not yet convinced this change is worth it.
> >
> > /cc #205 #226
> >
> > ―
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jdbi/jdbi/pull/235#issuecomment-160314040>.
>

qualidafial commented:
Heads up, I've been getting intermittent unit test timeouts since the last merge from jdbi3 branch.

brianm commented:
Definitely copy the class over if you want to make use of it, rather then depend on Guava,

qualidafial commented:
I've decided to migrate from Guava's TypeToken to Classmate's GenericType.

TypeToken is nice but it comes with a lot of baggage. Internally there's about a dozen classes that would have to come along, and it still doesn't compile in Java 8 since some new methods were added to java.lang.reflect element interfaces which Guava doesn't implement.

Classmate only does generic type reflection (which is all we really need).

Finally, I wasn't aware before yesterday that Classmate's `GenericType<T>` existed, which is roughly equivalent to Guava's `TypeToken<T>`. The primary reason I was using TypeToken was so we could have methods that made generics guarantees, e.g. that `ResultColumnMapperFactory.build(TypeToken<Foo>, ...)` returns a `ResultColumnMapper<? extends Foo>`. Since we can make the same guarantees with `GenericType<Foo>`, it's not necessary to add Guava as a dependency.

Once the Classmate migration is done, we can evaluate whether it's worthwhile to add it as an exported dependency, vs substituting our own version of GenericType while using a shaded Classmate internally.

qualidafial commented:
Migration to Classmate is done. Internally and in factory methods, ResolvedType is used. In API methods there is a Class<T> and a GenericType<T> method, and occasionally it was necessary to expose the ResolvedType method they delegate to.

qualidafial commented:
I've been reviewing this on and off for a week, and I think I'm now comfortable with where this PR is at.

Regarding the two outstanding questions in the PR description:

> Change SQLObject from Class to ~TypeToken~ GenericType?

I think we should hold off on this for now. Thus far I've seen no demand for a SQL object with generic types. At any rate, users have the workaround of declaring a subclass/subinterface which extends/implements the generic interface with a concrete parameter.

> Built in row mappers use ~TypeToken~ GenericType instead of Class?

The existing row mappers in question would be BeanMapper and JpaMapper. I think we should hold off on adding support for this speculatively. We could always add support for it in the future, using the ResolvedType scaffolding already in place.

This brings us to the question of Classmate. This has been an upstream dependency for some time now, however it has only been an internal, shaded dependency. There are compatibility and maintenance concerns if we decide to bless this library as an exported dependency.

Is the team comfortable with this?

qualidafial commented:
@brianm @stevenschlansker Any comments?

stevenschlansker commented:
Sorry, been buried in work.  Will look later this week.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/BuiltInArgumentFactory.java:
```diff
@@ -89,10 +94,25 @@ public Argument build(Class<?> expectedType, T value, StatementContext ctx)
             return new StringArgument(value.toString());
         }
 
+        if (value instanceof Optional) {
```
Having this here feels a bit icky

qualidafial replied:
Do you mean you'd like it in a separate factory?

stevenschlansker replied:
Maybe?  Not so convinced now that I think about it more.  I just hate the "chained special conditions" feeling of doing it first for Enum, then Optional, then ...

qualidafial replied:
It does disrupt the flow of the code. Maybe if I extract a private function?

qualidafial replied:
Have a look now.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/Foreman.java:
```diff
@@ -34,8 +37,15 @@
+
+    Argument waffle(ResolvedType expectedType, Object it, StatementContext ctx)
+    {
+        ResolvedType objectType = new TypeResolver().resolve(Object.class);
```
this is a constant?

qualidafial replied:
It could be.

One thing I've been meaning to ask you: what would be the value of reusing the same type resolver throughout, as in your PR? Is resolving types an expensive operation?

qualidafial replied:
Oh, I see your comment in the broader context. I just extracted this to a var for reuse in two different blocks, on lines 60 and 65 below.

stevenschlansker replied:
I don't know if there is a large efficiency gain or not.  The documentation encourages reuse of the `TypeResolver` instance so I was just blindly recommending doing so.  But yeah I really meant the `ResolvedType` here, not the `TypeResolver`.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/GeneratedKeys.java:
```diff
@@ -13,13 +13,14 @@
 import java.sql.Statement;
 
-import org.jdbi.v3.exceptions.ResultSetException;
-import org.jdbi.v3.tweak.ResultSetMapper;
```
lots of import shuffling, we should make sure our IDEs are configured with the same ordering...

qualidafial replied:
If you have a project code style file for e.g. IntelliJ or Eclipse I'd be happy to use it.

stevenschlansker replied:
Is there a good way to specify it for both?

This is what I have right now:

![screen shot 2015-12-10 at 4 57 17 pm](https://cloud.githubusercontent.com/assets/129097/11733098/2247228e-9f5f-11e5-97da-490d318c4f66.png)

stevenschlansker commented on core/src/main/java/org/jdbi/v3/SQLStatement.java:
```diff
@@ -1058,7 +1062,7 @@ public final SelfType bind(int position, Object value)
     public final SelfType bind(String name, Object value)
     {
-        return bind(name, getForeman().waffle(value != null ? value.getClass() : Object.class, value, getContext()));
+        return bind(name, getForeman().waffle(value == null ? Object.class : value.getClass(), value, getContext()));
```
why swap these?

qualidafial replied:
PMD has been drilling this one into me for years: with ternary and if/else statements, avoid `!` in the condition. It makes the flow easier to understand.

qualidafial replied:
See http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#ConfusingTernary

stevenschlansker commented on core/src/main/java/org/jdbi/v3/StatementContext.java:
```diff
@@ -61,7 +63,15 @@
+    <T> ResultColumnMapper<T> columnMapperFor(ResolvedType type);
+
+    /**
+     * Obtain an argument factory for given value in this context
```
Returns an argument, not an argument factory.

qualidafial replied:
Will fix.

stevenschlansker commented on core/src/test/java/org/jdbi/v3/TestMapArguments.java:
```diff
@@ -44,7 +44,7 @@ public void testNullBinding() throws Exception
         Map<String, Object> args = new HashMap<String, Object>();
         args.put("foo", null);
         Foreman foreman = new Foreman();
-        StatementContext ctx = new ConcreteStatementContext(new HashMap<String, Object>(), new MappingRegistry());
```
:+1:

stevenschlansker commented:
Generally LGTM.  Biggest hiccup is that it exposes Classmate as part of our public API, but I think that's a reasonable compromise.

Re: things like `collectInto`, we could consolidate it into `as(ResolvedType)` which does the equivalent of both `mapTo` and `collectInto` in one swoop?

qualidafial commented:
@stevenschlansker the `collectInto` vs `as` thing is a difficult problem.

I have some work in progress which adds a method `CollectorFactory.elementTypeFor(ResolvedType): ResolvedType` and uses that to map to the resolved type.

I'm implemented the map+collect feature. It works--sort of.

One problem is that it doesn't agree with the `@Mapper` annotation, which is a mapper override as opposed to a simple registration. I was able to work around this mismatch by only calling `this.mapTo` in the query when the current mapper is an instance of `DefaultMapper`, which is a major kludge.

Another issue is that it's just kind of confusing to use. You have the map and collect methods that exist now, but then you have this Frankenstein version of both of them. It creates odd use cases and ambiguous expectations. What happens if someone calls `query.mapTo(Foo.class).collectInto(new GenericType<Bar<Baz>>() {})`? Will it first map to `Foo`, then map to `Baz` and finally collect it into a `Bar`?

I'm still of the opinion that it's better to leave `mapTo(Class|GenericType)` in place alongside `collect(Collector<T, ?, R>) : R` which has pretty predictable behavior from a client perspective. I think we should only be talking to the collector registry in SQL objects. Then the SQL Objects API would be in control of when to call `mapTo(Foo.class)` vs `map(ResultSetMapper)` when e.g. there is a `@Mapper` annotation.

I'd prefer to separate this aspect out and look at it by itself in another branch, rather than hold up merge of this branch.

stevenschlansker commented:
Yes, we can certainly consider it separately.
Trying to see why the checks failed -- Travis just says "can't load the build data"

Do you think this branch is ready for merge?

brianm commented:
Today there are no hard external dependencies for JDBI. Is this worth creating one? Is there a reasonable way to accomplish this without adding an external lib to our public API?

FWIW, if we are going to open that door, I'd move to replace the entire mapping plumbing with Jackson, as we are not going to create something more flexible and powerful.

stevenschlansker commented:
It's really tough to work with the Java type system reflectively without some helper like Classmate.  The `Type` interface itself is essentially useless, and specializations like `ParameterizedType` are only marginally less useful.

We could consider just using the `Type` marker interface and expect each implementor to then deal with `Class` vs `ParameterizedType` with a strong encouragement to just defer to using Classmate.  That would probably lead to not exposing it via the API but have the cost of slightly more pain each time someone wants to implement a `*Factory` like interface.

qualidafial commented:
Based on this discussion I'll do a spike on expunging ResolvedType from our public APIs. I'll also replace `GenericType<T>` usage with a type token interface of our own.

This opens up the possibly to go back to a shaded Classmate API when we work out the problem with shading.

qualidafial commented:
I took a stab at this with Classmate. One issue that I haven't found a way around is that ResolvedType implements Type, but none of its subclasses implement standard `Type` subinterfaces i.e. `GenericArrayType`, `WildcardType`, `TypeVariable`, or `ParameterizedType`.

For this reason, I do not think it is feasible to use classmate internally for generic inspection, and still use `java.lang.reflect.Type` in a Classmate-agnostic way.

Guava's `TypeToken` does not share this deficiency. Guava does have its own classes for representing types, but they implement the `java.lang.reflect` classes.

qualidafial commented:
Only `sqlobject` uses Classmate now. Core and the rest are Guava- and Classmate-free.

Two types were added to core for dealing with generics: Types (a utility class) and GenericType<T> (similar to Classmate GenericType<T>). I was surprised to find that the only primitive generics operations we use are 1) get erased type, and 2) resolve the generic type parameter for a particular supertype.

brianm commented:
Awesome!

On Mon, Dec 28, 2015 at 12:41 AM, Matthew Hall <notifications@github.com>
wrote:

> Only sqlobject uses Classmate now. Core and the rest are Guava- and
> Classmate-free.
>
> Two types were added to core for dealing with generics: Types (a utility
> class) and GenericType (similar to Classmate GenericType). I was surprised
> to find that the only primitive generics operations we use are 1) get
> erased type, and 2) resolve the generic type parameter for a particular
> supertype.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jdbi/jdbi/pull/235#issuecomment-167513438>.
>

qualidafial commented:
Eh hem.. spoke too soon. With some additional testing, I've found my naive implementation of `Types.findGenericParameter()` is not resolving generics from superclasses. I'll have to reinstate the TypeToken stuff for the time being, with the upside being that everything is uses the `Types` class and nothing else is using Guava directly.

stevenschlansker commented on .travis.yml:
```diff
@@ -6,7 +6,7 @@ jdk:
 
 before_install:
-    - cp src/build/travis-toolchains.xml ~/.m2/toolchains.xml
+    - cp core/src/build/travis-toolchains.xml ~/.m2/toolchains.xml
```
I think this was fixed in #251 and this isn't needed?

qualidafial replied:
Yep, I'll revert.

stevenschlansker commented on core/src/build/travis-toolchains.xml:
```diff
@@ -0,0 +1,28 @@
+~   See the License for the specific language governing permissions and
+~   limitations under the License.
+-->
+<toolchains>
```
ditto, duplicate of #251

qualidafial commented on core/src/main/java/org/jdbi/v3/Query.java:
```diff
@@ -87,8 +89,18 @@
-        return collect(getCollectorFactoryRegistry().createCollectorFor(containerType));
+    public <R> R collectInto(GenericType<R> containerType) {
+        return this.<R>collectInto(containerType.getType());
+    }
```
I still think we should remove this method.

The `ResultType` parameter is erased at runtime, so we can't verify that `R` here is a valid signature for containing `ResultType`.

    handle.createQuery("select * from foo")
        .mapTo(Foo.class)
        .collectInto(new GenericType<ImmutableList<Bar>>() {});

This is a recipe for inexplicable `ClassCastException`s happening in client code after JDBI has returned.

@stevenschlansker I understand your desire to free the caller from having to know where the collector factories are, but I think the risk to type safety outweighs the convenience of this method.

Moreover, the code is actually shorter if you just use the `collect` method:

    handle.createQuery("select * from foo")
        .mapTo(Foo.class)
        .collect(GuavaCollectors.toImmutableList()); // <-- could be be statically imported too

stevenschlansker replied:
OK by me.

qualidafial commented on core/src/main/java/org/jdbi/v3/SQLStatement.java:
```diff
@@ -1140,16 +1141,59 @@ public final SelfType bind(String name, URL value)
+    }
+
+    /**
+     * Bind an argument dynamically by the type token passed in.
```
Change `type token` to `generic type`

qualidafial commented on core/src/main/java/org/jdbi/v3/SQLStatement.java:
```diff
@@ -1140,16 +1141,59 @@ public final SelfType bind(String name, URL value)
+    }
+
+    /**
+     * Bind an argument dynamically by the type token passed in.
```
Ditto

qualidafial commented on core/src/main/java/org/jdbi/v3/tweak/ResultColumnMapper.java:
```diff
@@ -20,7 +20,7 @@
 /**
  * Maps result set columns to objects.
- * @see StatementContext#columnMapperFor(Class)
+ * @see StatementContext#columnMapperFor(com.fasterxml.classmate.ResolvedType)
```
Change `ResolvedType` to `Type`

qualidafial commented on guava/src/main/java/org/jdbi/v3/guava/GuavaCollectors.java:
```diff
@@ -54,7 +57,7 @@
      * @return a collector into {@code ImmutableSortedSet<T>}
      */
-    public static <T extends Comparable<T>> Collector<T, ImmutableSortedSet.Builder<T>, ImmutableSortedSet<T>> toImmutableSortedSet() {
+    public static <T extends Comparable<T>> Collector<T, ?, ImmutableSortedSet<T>> toImmutableSortedSet() {
```
Unrelated, but for completeness we should include an additional method `toImmutableSortedSet(Comparator<T>)`

qualidafial replied:
:white_check_mark:

qualidafial commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/ResultReturnThing.java:
```diff
@@ -76,8 +77,8 @@ else if (return_type.isInstanceOf(Iterator.class)) {
-        private final Class<?> returnType;
-        private final Class<?> containerType;
+        private final ResolvedType returnType;
+        private final ResolvedType containerType;
```
Change to `Type`

qualidafial commented on sqlobject/src/main/java/org/jdbi/v3/sqlobject/ResultReturnThing.java:
```diff
@@ -238,24 +239,21 @@ public void remove()
-        private final ResolvedType resolvedType;
-        private final Class<?> erased_type;
+        private final ResolvedType iterableType;
+        private final ResolvedType elementType;
```
Change to `Type`

qualidafial commented on sqlobject/src/main/java/org/jdbi/v3/unstable/BindIn.java:
```diff
@@ -74,9 +76,17 @@ public SqlStatementCustomizer createForParameter(Annotation annotation,
             return (q, param, bind, arg) -> {
                 Iterable<?> coll = (Iterable<?>) arg;
+                ResolvedType elementType = new TypeResolver().resolve(Object.class);
+                ResolvedType parameterType = new TypeResolver().resolve(param.getParameterizedType());
```
Change to `Type`.

qualidafial commented on sqlobject/src/test/java/org/jdbi/v3/docs/TestInClauseExpansion.java:
```diff
@@ -63,18 +65,21 @@ public void testInClauseExpansion() throws Exception
-            return Collector.of(ImmutableSet.Builder::new, ImmutableSet.Builder::add, (first, second) -> {
-                throw new UnsupportedOperationException("Parallel collecting is not supported");
-            }, ImmutableSet.Builder::build, Collector.Characteristics.UNORDERED);
+        public Collector<T, ImmutableSet.Builder<T>, ImmutableSet<T>> newCollector(Type type) {
```
Accumulator parameter is unnecessary

qualidafial commented on sqlobject/src/test/java/org/jdbi/v3/docs/TestInClauseExpansion.java:
```diff
@@ -63,18 +65,21 @@ public void testInClauseExpansion() throws Exception
     }
 
-    public static class ImmutableSetCollectorFactory<T> implements CollectorFactory<T, ImmutableSet<T>> {
+    public static class ImmutableSetCollectorFactory<T> implements CollectorFactory {
```
Investigate whether this duplicate implementation of immutable set collector factory is necessary for this test. Remove if redundant.

qualidafial commented on sqlobject/src/test/java/org/jdbi/v3/docs/TestInClauseExpansion.java:
```diff
@@ -63,18 +65,21 @@ public void testInClauseExpansion() throws Exception
+                    ImmutableSet.Builder::add,
+                    (first, second) -> { throw new UnsupportedOperationException("Parallel collecting is not supported"); },
+                    ImmutableSet.Builder::build,
+                    Collector.Characteristics.UNORDERED);
```
If this implementation is integral to the test, just call `GuavaCollectors.immutableSet()` rather than reproducing the whole thing.

qualidafial commented on sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestCollectorFactory.java:
```diff
@@ -147,33 +147,37 @@ public void testWithSqlObjectSetReturnValue() throws Exception {
+                    ImmutableList.Builder::new,
+                    ImmutableList.Builder::add,
+                    (first, second) -> { throw new UnsupportedOperationException("Parallel collecting is not supported"); },
+                    ImmutableList.Builder<T>::build);
```
Call `GuavaCollectors.immutableList()` rather than duplicating code here.

qualidafial commented:
Could I get another review? I know this PR is getting huge..

stevenschlansker commented:
Yup, I'll look later today.  I really appreciate all the work on this, but yeah, reviewing it is getting tougher and tougher :)

qualidafial commented on core/src/main/java/org/jdbi/v3/BasicHandle.java:
```diff
@@ -242,30 +244,32 @@ public boolean isInTransaction()
                           timingCollector,
-                          foreman,
+                          updateForeman,
                           collectorFactoryRegistry);
```
Unrelated: Should this be calling `collectorFactoryRegistry.createChild()`?

stevenschlansker replied:
I think so, otherwise registering on the `Update` will cascade up to the `Handle`.

qualidafial replied:
Will fix

stevenschlansker commented on .gitignore:
```diff
@@ -19,3 +19,4 @@ out
 test-output
 dependency-reduced-pom.xml
 
+.factorypath
```
I don't see this referred to anywhere else, what is this?

qualidafial replied:
This is a project metadata file produced by eclipse.

stevenschlansker commented on core/pom.xml:
```diff
@@ -89,7 +84,6 @@
         <dependency>
             <groupId>com.google.guava</groupId>
             <artifactId>guava</artifactId>
-            <scope>test</scope>
```
I thought our goal was no Guava dependency -- are we giving up on that, or is this temporary?  Is there any hope of splitting off a `jdbi-guava` module instead of adding it to core?  (This will probably be answered by reading further, just noting things as I read top-to-bottom)

qualidafial replied:
Baby steps until either we get shaded jars working correctly (which I actually haven't seen any problems with in my local testing of maven-shade-plugin), or until we implement `Types.findGenericParameter` by hand.

qualidafial replied:
I added some additional tests and found that my by-hand implementation was naive--I switched back to TypeToken for correctness until I can come up with a correct hand-written implementation.

qualidafial replied:
Bear in mind that core only uses Guava internally--we are not exporting that dependency by including Guava types in our API signatures.

stevenschlansker replied:
OK then, that's fine by me.
re: shade plugin, I disabled it back when JDK8 was very new, and the shade plugin could not handle the new class file format yet.  I assume it works fine now and just needs testing.

However, I would push to have two artifacts -- maybe one without any classifier, and one with a `shaded` classifier?  At least for my projects I find shading in dependencies totally counterproductive, since we already have Guava and friends present, and that way we can pull in bugfixes with dependency management.  But if people really want shaded jars we can provide those too?

qualidafial replied:
That's what the plugin was doing by default IIRC

stevenschlansker commented on core/src/main/java/org/jdbi/v3/BasicHandle.java:
```diff
@@ -242,30 +244,32 @@ public boolean isInTransaction()
-                        foreman,
-                collectorFactoryRegistry);
+                        callForeman,
+                        collectorFactoryRegistry);
```
Ditto here?

qualidafial replied:
Yep

stevenschlansker commented on core/src/main/java/org/jdbi/v3/BasicHandle.java:
```diff
@@ -288,26 +292,28 @@ public int update(String sql, Object... args)
-                                 foreman,
-                collectorFactoryRegistry);
+                                 batchForeman,
+                                 collectorFactoryRegistry);
```
And here?

qualidafial replied:
Si

stevenschlansker commented:
OK, you've convinced me that we can ditch `collectInto`.  Make it so!

stevenschlansker commented on core/src/main/java/org/jdbi/v3/BuiltInArgumentFactory.java:
```diff
@@ -36,14 +41,14 @@
  * {@link SQLStatement} convenience methods.
  */
-public class BuiltInArgumentFactory<T> implements ArgumentFactory<T> {
+public class BuiltInArgumentFactory implements ArgumentFactory<Object> {
```
Yeah, this generics stuff is tricky.  I'm still not sure how we decide which things have a `<T>` or not in a principled manner.  I'm +1 to killing this one, though.

qualidafial replied:
In this case I didn't like how the `<T>` parameter just morphed into whatever we decided it was during the method call.

I actually think we should get rid of the parameter on `ArgumentFactory` like on the other factories--although I was planning to do that as part of the factory method coalescing I mentioned elsewhere on this ticket.

qualidafial replied:
To clarify: If the class is going to have a `T` parameter, than for the life of a particular generic object, that `T` should conceptually point to a single type. In this case we were using it as an end-run to save on casts.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/BuiltInArgumentFactory.java:
```diff
@@ -100,31 +105,54 @@ public static Argument build(Object arg) {
             return STR_BUILDER.build((((Enum<?>)value).name()));
         }
 
+        if (value instanceof Optional) {
```
Should this be handled here, or would a separate `OptionalArgumentFactory` be appropriate?  Feels like it is fundamentally different than any of the other types here, which are all simple mappings to `PreparedStatement#setXXX` methods.

qualidafial replied:
I was going by the "built in" part of the class name. There are certain things we'd expect to come built in and I would argue that Optional is one that people would expect out of the box.

qualidafial replied:
I guess maybe "built in" was referring to types recognized by `PreparedStatement` rather than by JDBI?

stevenschlansker commented on core/src/main/java/org/jdbi/v3/CollectorFactoryRegistry.java:
```diff
@@ -29,7 +32,7 @@
  */
 class CollectorFactoryRegistry {
```
So if we get rid of `collectInto` does all this `CollectorFactoryRegistry` stuff go away entirely?

qualidafial replied:
No, the registry is still needed by sqlobject.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/ConcreteStatementContext.java:
```diff
@@ -43,10 +46,16 @@
 
-    ConcreteStatementContext(Map<String, Object> globalAttributes, MappingRegistry mappingRegistry)
+    /* visible for testing */
+    ConcreteStatementContext() {
```
:+1:

stevenschlansker commented on core/src/main/java/org/jdbi/v3/ConcreteStatementContext.java:
```diff
@@ -89,11 +98,16 @@ public Object getAttribute(String key)
 
     @Override
-    public <T> ResultColumnMapper<T> columnMapperFor(Class<T> type)
+    public ResultColumnMapper<?> columnMapperFor(Type type)
```
Would the following method signatures be of interest for this sort of thing?
```java
public <T> ResultColumnMapper<T> columnMapperFor(Type type); // Not safe, but avoids a cast?
public <T> ResultColumnMapper<T> columnMapperFor(Class<T> type); // Keep the Type one, but add a more specific Class one to keep the type when we can?
```

qualidafial replied:
I'm ok with making type safety promises as long as there's a parameter to justify it:

    default <T> ResultColumnMapper<T> columnMapperFor(GenericType<T> genericType) {
      return (ResultColumnMapper<T>) columnMapperFor(genericType.getType());
    }
    default <T> ResultColumnMapper<T> columnMapperFor(Class<T> clazz) {
      return (ResultColumnMapper<T>) columnMapperFor((Type) clazz);
    }
    ResultColumnMapper<?> columnMapperFor(Type)

This means that factories are on the hook for type correctness--which is what we expect from them anyway.

qualidafial replied:
Can we hold off on this one? We can always add them as default methods. I'm not convinced we'll need them at the level of the library they'll be used.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/ConcreteStatementContext.java:
```diff
@@ -89,11 +98,16 @@ public Object getAttribute(String key)
 
+    @Override
+    public Argument argumentFor(Type type, Object value) {
+        return foreman.waffle(type, value, this);
```
We really need better names for these :)

qualidafial replied:
Eh hem.. argument registry?

stevenschlansker commented on core/src/main/java/org/jdbi/v3/InferredColumnMapperFactory.java:
```diff
@@ -13,42 +13,36 @@
-    public boolean accepts(Class<?> type, StatementContext ctx)
+    public boolean accepts(Type type, StatementContext ctx)
     {
         return maps.equals(type);
```
I'm not sure this is fixable, but would this be more correct with a `Type.isAssignableFrom(Type)` or something like that?  I worry that it is possible to have two equivalent types that do not compare equals, since this is an interface?

qualidafial replied:
TypeToken has an `isSuperTypeOf(Type)` method we could use.

It's hard to reason whether that's what we should do in this case though. If `SubtypeMapper` extends `SupertypeMapper`, do we want the subtype mapper to try to map all supertypes? I'm not sure that's appropriate.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/InferredColumnMapperFactory.java:
```diff
@@ -13,42 +13,36 @@
+    public ResultColumnMapper<?> columnMapperFor(Type type, StatementContext ctx)
     {
-        return (ResultColumnMapper<? extends T>) mapper;
+        if (!type.equals(maps)) {
```
Is this even possibly invoked?  Guess it doesn't harm anything to double check, but it sure makes the `accepts` method look useless if you have to double check here anyway.

qualidafial replied:
This was in `InferredResultSetMapperFactory`, I thought it was a good idea to copy it over.

I'm not sure how long it will be relevant though. Now that most factories have lost generic parameters, one of my next goals is to collapse the accept/build methods into a single method that possibly returns an Optional. At that point this check will be redundant anyway.

stevenschlansker commented on core/src/main/java/org/jdbi/v3/InferredMapperFactory.java:
```diff
@@ -13,44 +13,36 @@
-    InferredMapperFactory(ResultSetMapper<? extends X> mapper)
+    InferredMapperFactory(ResultSetMapper<X> mapper)
     {
+        this.maps = findGenericParameter(mapper.getClass(), ResultSetMapper.class)
```
I've seen this a couple of times, should we have a `findGenericParameterOrThrow` method?

qualidafial replied:
Optional already has this use case covered: `findGenericParameter(type, supertype).orElseThrow(() -> new FooException("error message"))`

stevenschlansker commented on core/src/main/java/org/jdbi/v3/MappingRegistry.java:
```diff
@@ -26,10 +27,10 @@
 
     private final List<ResultSetMapperFactory> rowFactories = new CopyOnWriteArrayList<>();
-    private final ConcurrentHashMap<Class<?>, ResultSetMapper<?>> rowCache = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<Type, ResultSetMapper<?>> rowCache = new ConcurrentHashMap<>();
```
I worry a little bit about a `Map` with an key that is an interface that doesn't have a specified equality.  Maybe I worry too much...

stevenschlansker replied:
Looks like at least `ParameterizedType` does okay: http://hg.openjdk.java.net/jdk7/jdk7/jdk/file/572791538be5/src/share/classes/sun/reflect/generics/reflectiveObjects/ParameterizedTypeImpl.java#l155

qualidafial replied:
From what I can see, all of the types included in the SDK implement equals using the public API of each type interface. Guava does this as well.

qualidafial replied:
FWIW `TypeToken.equals` relies on `Type.equals`

stevenschlansker commented on core/src/main/java/org/jdbi/v3/StatementContext.java:
```diff
@@ -61,7 +63,15 @@
+     * Obtain an argument for given value in this context
+     * @param type the type of the argument.
+     * @param value the argument value.
+     * @return an Argument for the given value, or null if no argument factory is registered for the given type
```
Will this actually ever return null?  `BuiltInArgumentFactory` will in the worst case go with `PreparedStatement.setObject`...

qualidafial replied:
Nope, it won't return null. I will amend the javadoc.

stevenschlansker commented on core/src/test/java/org/jdbi/v3/GenericTypeTest.java:
```diff
@@ -0,0 +1,34 @@
+
+public class GenericTypeTest {
+
+    @Test
```
Probably worth a few more tests :smile:

qualidafial replied:
Will do

stevenschlansker commented on core/src/test/java/org/jdbi/v3/TestOptional.java:
```diff
@@ -0,0 +1,177 @@
+    public void testDynamicBindOptionalOfUnregisteredCustomType() throws Exception {
+        exception.expect(UnableToCreateStatementException.class);
+        handle.createQuery(SELECT_BY_NAME)
+                .dynamicBind(new GenericType<Optional<Name>>() {}, "name", Optional.of(new Name("eric")))
```
Nice tests here

stevenschlansker commented on core/src/test/java/org/jdbi/v3/TypesTest.java:
```diff
@@ -0,0 +1,105 @@
+
+import org.junit.Test;
+
+public class TypesTest {
```
Aaah, here are the more thorough generic tests.

stevenschlansker commented on core/src/test/java/org/jdbi/v3/TypesTest.java:
```diff
@@ -0,0 +1,105 @@
+    private Type methodReturnType(String methodName) throws NoSuchMethodException {
+        return getClass().getDeclaredMethod(methodName).getGenericReturnType();
+    }
+
```
Maybe worth testing some mismatched types e.g. `Baz<R> extends Bar<String>` and some wildcard types, including `? super T` and the like?

qualidafial replied:
Will do. If you could suggest some specific scenarios that might help--I've gone a little snow blind trying to reason through these type tests.

stevenschlansker commented on guava/src/main/java/org/jdbi/v3/guava/GuavaCollectors.java:
```diff
@@ -32,62 +38,103 @@
+                ImmutableSortedSet.Builder::build);
+    }
+
+    private static <T, C extends ImmutableCollection.Builder<T>> C combineBuilders(C left, C right) {
```
Nice!

stevenschlansker commented on jpa/src/main/java/org/jdbi/v3/jpa/EntityMemberAccessException.java:
```diff
@@ -15,6 +15,7 @@
 
 import org.jdbi.v3.exceptions.DBIException;
 
+@SuppressWarnings("serial")
```
I usually just add the `serialVersionUID` field rather than suppressing the warning.

qualidafial replied:
Will do

stevenschlansker commented:
Lookin' good!  Unfortunately I'm gone for a week and a half to Taiwan so I won't be following up soon, but I left some comments.

qualidafial commented:
Thanks Steven. Enjoy your trip!

qualidafial commented:
@stevenschlansker Are you comfortable with me merging this PR once I've addressed your comments?

qualidafial commented:
@stevenschlansker @brianm I believe I've addressed all the issues that are critical for merge. The rest I added to `CHANGES_PLANNED_FOR_3_0` so they don't get dropped.

+1?

stevenschlansker commented:
Thanks for your continued work here and patience with getting it merged.  Any further problems we can fix in post.  :shipit:

stevenschlansker commented:
Thanks for your continued work here and patience with getting it merged.  Any further problems we can fix in post.  :shipit:
